lucene/highlighter/src/test/org/apache/lucene/search/matchhighlight/TestMatchHighlighter.java

```diff
@@ -0,0 +1,466 @@
+    // glued together.
+    final int offsetGap = RandomizedTest.randomIntBetween(1, 2);
+    final int positionGap = RandomizedTest.randomFrom(new int[]{0, 1, 100});
+    Analyzer synonymsAnalyzer =
+        new AnalyzerWithGaps(offsetGap, positionGap, new Analyzer() {
+          @Override
+          protected TokenStreamComponents createComponents(String fieldName) {
+            Tokenizer tokenizer = new WhitespaceTokenizer();
+            TokenStream tokenStream = new SynonymGraphFilter(tokenizer, synonymMap, true);
+            return new TokenStreamComponents(tokenizer, tokenStream);
+          }
+        });
+
+    fieldAnalyzers.put(FLD_TEXT1, synonymsAnalyzer);
+    fieldAnalyzers.put(FLD_TEXT2, synonymsAnalyzer);
+
+    analyzer = new PerFieldAnalyzerWrapper(new MissingAnalyzer(), fieldAnalyzers);
+  }
+
+  static SynonymMap buildSynonymMap(String[][] synonyms) throws IOException {
+    SynonymMap.Builder builder = new SynonymMap.Builder();
+    for (String[] pair : synonyms) {
+      assertThat(pair.length, Matchers.equalTo(2));
+      builder.add(new CharsRef(pair[0]), new CharsRef(pair[1]), true);
+    }
+    return builder.build();
+  }
+
+  @Test
+  public void testBasicUsage() throws IOException {
+    new IndexBuilder(this::toField)
+        .doc(FLD_TEXT1, "foo bar baz")
+        .doc(FLD_TEXT1, "bar foo baz")
+        .doc(fields -> {
+          fields.add(FLD_TEXT1, "Very long content but not matching anything.");
+          fields.add(FLD_TEXT2, "no foo but bar");
+        })
+        .build(analyzer, reader -> {
+          Query query = new BooleanQuery.Builder()
+              .add(new TermQuery(new Term(FLD_TEXT1, "foo")), BooleanClause.Occur.SHOULD)
+              .add(new TermQuery(new Term(FLD_TEXT2, "bar")), BooleanClause.Occur.SHOULD)
+              .build();
+
+          // In the most basic scenario, we run a search against a query, retrieve
+          // top docs...
+          IndexSearcher searcher = new IndexSearcher(reader);
+          Sort sortOrder = Sort.INDEXORDER; // So that results are consistently ordered.
+          TopDocs topDocs = searcher.search(query, 10, sortOrder);
+
+          // ...and would want a fixed set of fields from those documents, some of them
+          // possibly highlighted if they matched the query.
+          //
+          // This configures the highlighter so that the FLD_ID field is always returned verbatim,
+          // and FLD_TEXT1 is returned *only if it contained a query match*.
+          MatchHighlighter highlighter =
+              new MatchHighlighter(searcher, analyzer)
+                .appendFieldHighlighter(FieldValueHighlighters.verbatimValue(FLD_ID))
+                .appendFieldHighlighter(FieldValueHighlighters.highlighted(
+                    80 * 3, 1, new PassageFormatter("...", ">", "<"), FLD_TEXT1::equals))
+                .appendFieldHighlighter(FieldValueHighlighters.skipRemaining());
```

Comment:
Just out of interest, what happens if you don't add a `skipRemaining()` highlighter to the end?  If it's necessary, maybe we should convert to a Builder pattern and automatically add it to the end of the highlighters list before returning the MatchHighlighter?

Reply:
These field highlighters can apply to more than one field and the highlighter may receive fields it wasn't prepared to process (because they were part of the query). I opted to be explicit - if no field highlighter "accepts" a field, the entire highlighting process fails. The "skip remaining" is essentially a rule that does nothing for any field that wasn't consumed before.

Performance-wise I admit I wasn't too keen on optimizing things too early. This thing, even if slower, just works for me in 
so many more scenarios than any of the approaches I've tried with other highlighters in Lucene. That "two queries at once" example isn't entirely out of thin air... we actually do use something like this to highlight queries originating from different sources.